Adding the QueueDb and using it in background sync instead of DBWrapper

tropicadri opened the pull request:
This is the first part of #2801 
I created a new class to interact with IDB, to manage the queues in workbox-background-sync. This new class encapsulates details like the object store name and index, as well as the types the DB should manage.

tropicadri committed:
Adding the QueueDb and using it in background sync instead of DBWrapper

jeffposnick commented on packages/workbox-background-sync/src/Queue.ts:
```diff
@@ -10,11 +10,11 @@ import {WorkboxError} from 'workbox-core/_private/WorkboxError.js';
 import {getFriendlyURL} from 'workbox-core/_private/getFriendlyURL.js';
-import {UnidentifiedQueueStoreEntry, QueueStore} from './lib/QueueStore.js';
+import {QueueStore} from './lib/QueueStore.js';
+import {UnidentifiedQueueStoreEntry} from './lib/QueueDb';
```
This needs to have a `.js` file extension or certain use cases break.

tropicadri replied:
updated.

jeffposnick commented on package.json:
```diff
@@ -81,6 +81,7 @@
     "html-webpack-plugin-v4": "npm:html-webpack-plugin@^4.5.0",
     "html-webpack-plugin-v5": "npm:html-webpack-plugin@^5.0.0-alpha.14",
     "husky": "^4.2.5",
+    "idb": "^6.0.0",
```
Just double-checking—what necessitates adding in the `idb` dependency at the top-level?Is it there for the test server, for creating our `build/` bundles, or some other purpose?

tropicadri replied:
For the test server. Correct me if I'm wrong but to use it in the tests, it needs to be included at the top-level. That is why I added the 'node_modules' as a directory in sw-bundle.js. Not sure if there is another way to do it.

jeffposnick replied:
👍

jeffposnick commented on packages/workbox-background-sync/src/lib/QueueDb.ts:
```diff
@@ -0,0 +1,149 @@
+/*
+  Copyright 2018 Google LLC
```
You can switch this to 2021.

jeffposnick commented on packages/workbox-background-sync/src/lib/QueueDb.ts:
```diff
@@ -0,0 +1,149 @@
+  timestamp: number;
+  id?: number;
+  queueName?: string;
+  metadata?: object;
```
I know that this is what the old TypeScript definition used, but maybe there's an opportunity to clean that up a bit to remove some of the ambiguity, since `object` is overly broad: https://2ality.com/2020/01/typing-objects-typescript.html#object-vs.-object-in-typescript

Something like `Record<string, any>` might make more sense, or something more strict if we know that the values of each property will only be a certain subset of types.

(I'm assuming we can making "breaking" changes like that for this definition since it's internal-only, and not part of the public API, but correct me if I'm wrong.)

tropicadri replied:
We talked about this a bit, but we ultimately expose it as public as part of the API: https://developers.google.com/web/tools/workbox/reference-docs/latest/module-workbox-background-sync.Queue#pushRequest so won't change it for now.

jeffposnick replied:
👍

jeffposnick commented on packages/workbox-background-sync/src/lib/QueueDb.ts:
```diff
@@ -0,0 +1,149 @@
+}
+
+/**
+ * A class to interact directly with IndexedDB to store a retrieve queue requests.
```
I think there's a typo in this description.

jeffposnick commented on packages/workbox-background-sync/src/lib/QueueDb.ts:
```diff
@@ -0,0 +1,149 @@
+  }
+
+  /**
+   *
```
If this is meant to be public, could you add a JSDoc description of what it does? Otherwise, add `@private`.

(That applies to the other methods as well.)

jeffposnick commented on packages/workbox-background-sync/src/lib/QueueDb.ts:
```diff
@@ -0,0 +1,149 @@
+   *
+   * @private
+   */
+  private async open() {
```
Usage might be cleaner if you have this method always return `this._db`, and name it `getDB()` or something like that.

That way you wouldn't have to put the `!` after `this._db!` all over the place.

tropicadri replied:
Agreed, fixed it.

jeffposnick commented on packages/workbox-background-sync/src/lib/QueueDb.ts:
```diff
@@ -0,0 +1,149 @@
+   *
+   * @return {QueueStoreEntry}
+   */
+  async getFirstEntry(): Promise<QueueStoreEntry | any> {
```
`Promise<QueueStoreEntry | any>` is the same as `Promise<any>` (for the same reason that `someValue || true` is the same as `true` in logic).

I think you mentioned that this might be `Promise<QueueStoreEntry | undefined>`?

tropicadri replied:
Yes, I fixed them.

jeffposnick commented on packages/workbox-background-sync/src/lib/QueueDb.ts:
```diff
@@ -0,0 +1,149 @@
+      .store.openCursor();
+
+    if (cursor) {
+      return cursor.value as QueueStoreEntry;
```
I'm kind of surprised that you have to cast `cursor.value` to `QueueStoreEntry` here—what is the type of `cursor.value` if you don't cast? One of the advantages of the IDB library, I thought, was that it "knew" about the types defined in `QueueDBSchema`.

tropicadri replied:
There are two interfaces defined for an entry UnidentifiedQueueStoreEntry and QueueStoreEntry, the difference is that QueueStoreEntry has to have an id, for the UnidentifiedQueueStoreEntry the id is optional. 

To add a record to it has to be unidentified because IDB sets the for the record when it. So the type is specified in the schema as UnidentifiedQueueEntry, line 16. The methods return Promise<UnidentifiedQueryEntry> that is why the need for the casting. No sure if there is a better way to handle that.

jeffposnick commented on packages/workbox-background-sync/src/lib/QueueDb.ts:
```diff
@@ -0,0 +1,149 @@
+  ): Promise<QueueStoreEntry[]> {
+    await this.open();
+    const results =
+      await this._db?.getAllFromIndex(OBJECT_STORE_NAME, INDEXED_PROP, query) as QueueStoreEntry[];
```
As with previous comment, is casting to `QueueStoreEntry[]` necessary? If `getAllFromIndex()` is returning something really generic, like `any[]` by default, then maybe the IDB library's schema definition isn't being set properly?

tropicadri replied:
See above.

jeffposnick commented on packages/workbox-background-sync/src/lib/QueueDb.ts:
```diff
@@ -0,0 +1,149 @@
+  async getEndEntryFromIndex(
+    {direction}: {direction?: IDBCursorDirection},
+    query?: IDBKeyRange,
+  ): Promise<QueueStoreEntry | any> {
```
Same comment about `Promise<QueueStoreEntry | any>` being the same as `Promise<any>`.

jeffposnick commented on packages/workbox-background-sync/src/lib/QueueDb.ts:
```diff
@@ -0,0 +1,149 @@
+   * @return {Promise<QueueStoreEntry | any>}
+   */
+  async getEndEntryFromIndex(
+    {direction}: {direction?: IDBCursorDirection},
```
It seems a little awkward to have one parameter as a destructured object, and the second parameter as just a normal object.

Since I think this is not exposed publicly, do you want to just change the method signature so that it's either a single parameter that gets destructured, or alternatively two parameters without any destructuring? So:

```
async getEndEntryFromIndex({direction, query}: {
  direction?: IDBCursorDirection,
  query?: IDBKeyRange,
})
```

or

```
async getEndEntryFromIndex(direction?: IDBCursorDirection, query?: IDBKeyRange)
```

Also, if you know that one or both of the parameter values will always be set, you don't need to add a `?` after them. It's a TypeScript best practice to be as strict as possible and only loosen up restrictions (via `?`) if there are times when you really don't know if the value will be `undefined`.

tropicadri replied:
Yes, copy pasted from QueueStore, restructured all this for better readability + clearer names.

jeffposnick commented on packages/workbox-background-sync/src/lib/QueueStore.ts:
```diff
@@ -7,27 +7,8 @@
-export interface QueueStoreEntry extends UnidentifiedQueueStoreEntry {
-  id: number;
-}
+import {UnidentifiedQueueStoreEntry, QueueStoreEntry, QueueDb} from './QueueDb';
```
Please add the `.js` suffix.

tropicadri replied:
We should find a way to add this to our lint :)

jeffposnick commented on packages/workbox-background-sync/src/lib/QueueStore.ts:
```diff
@@ -153,11 +130,10 @@ export class QueueStore {
-      index: INDEXED_PROP,
-      query: IDBKeyRange.only(this._queueName),
-    });
+  async getAll(): Promise<QueueStoreEntry[] | any> {
```
Same comment about how `Promise<QueueStoreEntry[] | any>` really means `Promise<any>`.

tropicadri replied:
Fixed.

jeffposnick commented on packages/workbox-background-sync/src/lib/QueueStore.ts:
```diff
@@ -183,38 +159,14 @@ export class QueueStore {
-      query: IDBKeyRange.only(this._queueName),
-      count: 1,
-    });
+    const entry = await this._queueDb.getEndEntryFromIndex(
```
See my earlier comment about the method signature for `getEndEntryFromIndex()`, as passing in the first parameter wrapped in `{}` looks awkward.

jeffposnick reviewed (changes requested):
Thanks for working on this!

In addition to the inline comments, I had a question about `test-QueueStore.mjs`. There are a lot of test cases removed, but no equivalents added. Is this because the test cases cover usage of `idb` that is already tested by the library's own test suite?

tropicadri committed:
updating the background sync package-lock.json with the idb dependency

tropicadri committed:
adding _version.js import

tropicadri committed:
fixing version import

tropicadri committed:
Improving documentation and chaging QueueDb interface to handle everything related to IDB

tropicadri commented:
> Thanks for working on this!
> 
> In addition to the inline comments, I had a question about `test-QueueStore.mjs`. There are a lot of test cases removed, but no equivalents added. Is this because the test cases cover usage of `idb` that is already tested by the library's own test suite?

Not sure if you saw this file test-QueueDb.mjs: https://github.com/GoogleChrome/workbox/pull/2805/files#diff-0527463f9abf935dcbb6cd2edf286adef83a169bbccfa88aa1bc43e18518cd35 (github collapses the file because the diff is big)

The tests that I removed from test-QueueStore.mjs where moved there, because the functionality is now implemented there + added more tests for the class

jeffposnick commented:
I have been burned so many times by GitHub collapsing files that have too many changes. (**The files with a lot of changes are the ones I want to pay closest attention to, GitHub!**)

I'll take a look at the new tests.

jeffposnick reviewed (changes requested):
The tests that I originally overlooked look good!

The main unresolved issue has to do with the distinction between `UnidentifiedQueueStoreEntry` and `QueueStoreEntry`.

Not surprisingly, other folks have run into this same issue, it seems: https://github.com/jakearchibald/idb/issues/150

Until that underlying `idb` library issue is fixed, it sounds like we need to work around it in our code. I've got an inline comment about doing the type conversion inside of `addEntry()` instead of when IDB is read from—I think that would be cleaner overall.

jeffposnick commented on packages/workbox-background-sync/src/lib/QueueDb.ts:
```diff
@@ -0,0 +1,167 @@
+interface QueueDBSchema extends DBSchema {
+  requests: {
+    key: number;
+    value: UnidentifiedQueueStoreEntry;
```
Can this be switched to `QueueStoreEntry`?

jeffposnick commented on packages/workbox-background-sync/src/lib/QueueDb.ts:
```diff
@@ -0,0 +1,167 @@
+   */
+  async addEntry(entry: UnidentifiedQueueStoreEntry) {
+    const db = await this.getDb();
+    await db.add(REQUEST_OBJECT_STORE_NAME, entry);
```
Can this be switched to

```
await db.add(REQUEST_OBJECT_STORE_NAME, entry as QueueStoreEntry);
```

I think if we just "fake" the type info in that one place, and everywhere that we read from IDB we get back a `QueueStoreEntry`, that would be cleaner overall.

jeffposnick commented on packages/workbox-background-sync/src/lib/QueueDb.ts:
```diff
@@ -0,0 +1,167 @@
+        upgrade: this._upgradeDb,
+      });
+    }
+    return this._db
```
Missing trailing `;`.

(I don't know why our linting doesn't pick this up. Maybe once https://github.com/GoogleChrome/workbox/issues/2602 is resolved it might?)

jeffposnick commented on packages/workbox-background-sync/src/lib/QueueDb.ts:
```diff
@@ -0,0 +1,167 @@
+    const cursor = await db.transaction(REQUEST_OBJECT_STORE_NAME)
+      .store.index(QUEUE_NAME_INDEX)
+      .openCursor(query, direction);
+    return cursor?.value as QueueStoreEntry;
```
If you change the schema to define `value` as `QueueStoreEntry` you shouldn't have to cast here anymore.

jeffposnick commented on packages/workbox-background-sync/src/lib/QueueDb.ts:
```diff
@@ -0,0 +1,167 @@
+    queueName: string,
+  ): Promise<QueueStoreEntry[]> {
+    const db = await this.getDb();
+    const results = await db.getAllFromIndex(REQUEST_OBJECT_STORE_NAME, QUEUE_NAME_INDEX, IDBKeyRange.only(queueName)) as QueueStoreEntry[];
```
You can remove this type conversion if you change the `value` field to be of type `QueueStoreEntry`.

jeffposnick reviewed (changes requested):
(Sorry, these are the comments that were meant to be attached to the previous review.)

The tests that I originally overlooked look good!

The main unresolved issue has to do with the distinction between `UnidentifiedQueueStoreEntry` and `QueueStoreEntry`.

Not surprisingly, other folks have run into this same issue, it seems: https://github.com/jakearchibald/idb/issues/150

Until that underlying `idb` library issue is fixed, it sounds like we need to work around it in our code. I've got an inline comment about doing the type conversion inside of `addEntry()` instead of when IDB is read from—I think that would be cleaner overall.

tropicadri committed:
Merge branch 'v6' into dbwrapper-deprecation

tropicadri committed:
Modifiying DB types to keep code cleaner

jeffposnick reviewed (approved):
Awesome! Looks good to me.

tropicadri committed:
Merge branch 'v6' into dbwrapper-deprecation

tropicadri committed:
Update package.json
